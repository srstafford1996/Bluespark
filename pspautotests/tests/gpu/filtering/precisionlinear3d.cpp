#include "shared.h"

extern "C" int HAS_DISPLAY;

typedef struct {
	float u, v;
	u32 color;
	float x, y, z;
} VertexColorF32;

static __attribute__((aligned(16))) VertexColorF32 vertices_f32[256];
static __attribute__((aligned(16))) u32 texdata[16] = {
	0xFFFFFFFF, 0xFF000000, 0xFFFFFFFF, 0xFFFFFFFF,
	0xFF0000FF, 0xFF00FF00, 0xFFFFFFFF, 0xFFFFFFFF,
	0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
	0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF,
};

inline VertexColorF32 makeVertex32(float u, float v, u32 c, float x, float y, float z) {
	VertexColorF32 vv;
	vv.u = u;
	vv.v = v;
	vv.color = c;
	vv.x = x;
	vv.y = y;
	vv.z = z;
	return vv;
}

void displayBuffer(const char *reason) {
	u32 tl = readFullDispBuffer(0, 0);

	u32 red = tl;
	int redpos = -1;
	for (int y = 1; y < 272; ++y) {
		u32 c = readFullDispBuffer(0, y);
		if (c != tl) {
			red = c;
			redpos = y;
			break;
		}
	}

	u32 black = tl;
	int blackpos = -1;
	for (int x = 1; x < 480; ++x) {
		u32 c = readFullDispBuffer(x, 0);
		if (c != tl) {
			black = c;
			blackpos = x;
			break;
		}
	}

	checkpoint("%s: COLOR=%08x, left=%d, %08x, top=%d, %08x", reason, tl, redpos, red, blackpos, black);

	// Reset.
	clearDispBuffer(0x44444444);
}

void init() {
	initDisplay();

	setNeedFull(true);
	clearDispBuffer(0x44444444);
	displayBuffer("Initial");

	startFrame();
	sceGuTexFilter(GU_LINEAR, GU_LINEAR);
	sceGuTexImage(0, 2, 2, 4, texdata);
	sceGuTexWrap(GU_CLAMP, GU_CLAMP);
	// Let's simply scale the UVs and pos to match through.
	sceGuTexScale(0.5f, 0.5f);
	sceGuOffset(2048, 2048);
	sceGuViewport(2048, 2048, 2, -2);
	sceGuSendCommandf(68, 1.0f);
	sceGuSendCommandf(71, 0.0f);
	sceGuTexFlush();
	sceGuTexSync();
	endFrame();
}

void testTexMagnify(const char *title, int w, int h, float xoff, float yoff) {
	startFrame();
	if (w >= 0 && h >= 0) {
		vertices_f32[0] = makeVertex32(0.0f + xoff, 0.0f + yoff, 0xFFFFFFFF, 0.0f, 0.0f, 0.0f);
		vertices_f32[1] = makeVertex32(2.0f + xoff, 2.0f + yoff, 0xFFFFFFFF, w, h, 0.0f);
	} else if (w >= 0) {
		vertices_f32[0] = makeVertex32(0.0f + xoff, 0.0f + yoff, 0xFFFFFFFF, 0.0f, -h, 0.0f);
		vertices_f32[1] = makeVertex32(2.0f + xoff, 2.0f + yoff, 0xFFFFFFFF, w, 0.0f, 0.0f);
	} else if (h >= 0) {
		vertices_f32[0] = makeVertex32(0.0f + xoff, 0.0f + yoff, 0xFFFFFFFF, -w, 0.0f, 0.0f);
		vertices_f32[1] = makeVertex32(2.0f + xoff, 2.0f + yoff, 0xFFFFFFFF, 0.0f, h, 0.0f);
	} else {
		vertices_f32[0] = makeVertex32(0.0f + xoff, 0.0f + yoff, 0xFFFFFFFF, -w, -h, 0.0f);
		vertices_f32[1] = makeVertex32(2.0f + xoff, 2.0f + yoff, 0xFFFFFFFF, 0.0f, 0.0f, 0.0f);
	}

	// Clearing cache is fun.  Let's do it all the time.
	sceKernelDcacheWritebackInvalidateRange(vertices_f32, sizeof(vertices_f32));
	sceGuDrawArray(GU_SPRITES, GU_TEXTURE_32BITF | GU_COLOR_8888 | GU_VERTEX_32BITF | GU_TRANSFORM_3D, 2, NULL, vertices_f32);
	dirtyDispBuffer();
	endFrame();

	displayBuffer(title);
}

extern "C" int main(int argc, char *argv[]) {
	init();
	HAS_DISPLAY = 0;

	checkpointNext("Orientations:");
	testTexMagnify("  TL -> BR", 256.0f, 256.0f, 0.0f, 0.0f);
	testTexMagnify("  BR -> TL", -256.0f, -256.0f, 0.0f, 0.0f);
	testTexMagnify("  TR -> BL", -256.0f, 256.0f, 0.0f, 0.0f);
	testTexMagnify("  BL -> TR", 256.0f, -256.0f, 0.0f, 0.0f);
	testTexMagnify("  TL -> BR (small)", 2.0f, 2.0f, 0.0f, 0.0f);

	checkpointNext("Offsets");
	testTexMagnify("  One texel", 256.0f, 256.0f, 1.0f, 1.0f);
	testTexMagnify("  Half texel", 256.0f, 256.0f, 0.5f, 0.5f);
	testTexMagnify("  Negative half texel", 256.0f, 256.0f, -0.5f, -0.5f);
	testTexMagnify("  Half texel (small)", 2.0f, 2.0f, 0.5f, 0.5f);

	checkpointNext("Pos X offsets:");
	for (int i = 0; i < 16; ++i) {
		sceGuStart(GU_DIRECT, list);
		sceGuSendCommandi(76, 2048 * 16 + i);
		sceGuSendCommandi(77, 2048 * 16);
		sceGuFinish();
		sceGuSync(GU_SYNC_WAIT, GU_SYNC_WHAT_DONE);

		char temp[256];
		snprintf(temp, sizeof(temp), "  Pos X offset %d", i);
		testTexMagnify(temp, 2.0f, 2.0f, 0.0f, 0.0f);
		snprintf(temp, sizeof(temp), "  Pos X offset (256x) %d", i);
		testTexMagnify(temp, 512.0f, 512.0f, 0.0f, 0.0f);
	}

	checkpointNext("Pos Y offsets:");
	for (int i = 0; i < 16; ++i) {
		sceGuStart(GU_DIRECT, list);
		sceGuSendCommandi(76, 2048 * 16);
		sceGuSendCommandi(77, 2048 * 16 + i);
		sceGuFinish();
		sceGuSync(GU_SYNC_WAIT, GU_SYNC_WHAT_DONE);

		char temp[256];
		snprintf(temp, sizeof(temp), "  Pos Y offset %d", i);
		testTexMagnify(temp, 2.0f, 2.0f, 0.0f, 0.0f);
		snprintf(temp, sizeof(temp), "  Pos Y offset (256x) %d", i);
		testTexMagnify(temp, 512.0f, 512.0f, 0.0f, 0.0f);
	}

	sceGuTerm();

	return 0;
}
