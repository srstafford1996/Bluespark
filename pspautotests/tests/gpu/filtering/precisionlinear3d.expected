[x] Initial: COLOR=44444444, left=-1, 44444444, top=-1, 44444444

[x] Orientations:
[r]   TL -> BR: COLOR=44ffffff, left=72, 44efefff, top=72, 44efefef
[r]   BR -> TL: COLOR=4400ff00, left=64, 4400ef00, top=64, 4400ef0f
[r]   TR -> BL: COLOR=440000ff, left=72, 44000fef, top=64, 440f0fff
[r]   BL -> TR: COLOR=44000000, left=64, 440f0f0f, top=72, 44000f00
[r]   TL -> BR (small): COLOR=44ffffff, left=1, 440000ff, top=1, 44000000

[x] Offsets
[r]   One texel: COLOR=443f7f7f, left=8, 44377f7f, top=8, 44377f6f
[r]   Half texel: COLOR=44ffffff, left=8, 44efefff, top=8, 44efefef
[r]   Negative half texel: COLOR=44ffffff, left=136, 44efefff, top=136, 44efefef
[r]   Half texel (small): COLOR=443f7f7f, left=1, 44007f7f, top=1, 44007f00

[x] Pos X offsets:
[r]   Pos X offset 0: COLOR=44ffffff, left=1, 440000ff, top=1, 44000000
[r]   Pos X offset (256x) 0: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos X offset 1: COLOR=44efefef, left=1, 44000fef, top=1, 44000000
[r]   Pos X offset (256x) 1: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos X offset 2: COLOR=44dfdfdf, left=1, 44001fdf, top=1, 44000000
[r]   Pos X offset (256x) 2: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos X offset 3: COLOR=44cfcfcf, left=1, 44002fcf, top=1, 44000000
[r]   Pos X offset (256x) 3: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos X offset 4: COLOR=44bfbfbf, left=1, 44003fbf, top=1, 44000000
[r]   Pos X offset (256x) 4: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos X offset 5: COLOR=44afafaf, left=1, 44004faf, top=1, 44000000
[r]   Pos X offset (256x) 5: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos X offset 6: COLOR=449f9f9f, left=1, 44005f9f, top=1, 44000000
[r]   Pos X offset (256x) 6: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos X offset 7: COLOR=448f8f8f, left=1, 44006f8f, top=1, 44000000
[r]   Pos X offset (256x) 7: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos X offset 8: COLOR=447f7f7f, left=1, 44007f7f, top=1, 44444444
[r]   Pos X offset (256x) 8: COLOR=44ffffff, left=144, 44efefff, top=143, 44efefef
[r]   Pos X offset 9: COLOR=446f6f6f, left=1, 44008f6f, top=1, 44444444
[r]   Pos X offset (256x) 9: COLOR=44ffffff, left=144, 44efefff, top=143, 44efefef
[r]   Pos X offset 10: COLOR=445f5f5f, left=1, 44009f5f, top=1, 44444444
[r]   Pos X offset (256x) 10: COLOR=44ffffff, left=144, 44efefff, top=143, 44efefef
[r]   Pos X offset 11: COLOR=444f4f4f, left=1, 4400af4f, top=1, 44444444
[r]   Pos X offset (256x) 11: COLOR=44ffffff, left=144, 44efefff, top=143, 44efefef
[r]   Pos X offset 12: COLOR=443f3f3f, left=1, 4400bf3f, top=1, 44444444
[r]   Pos X offset (256x) 12: COLOR=44ffffff, left=144, 44efefff, top=143, 44efefef
[r]   Pos X offset 13: COLOR=442f2f2f, left=1, 4400cf2f, top=1, 44444444
[r]   Pos X offset (256x) 13: COLOR=44ffffff, left=144, 44efefff, top=143, 44efefef
[r]   Pos X offset 14: COLOR=441f1f1f, left=1, 4400df1f, top=1, 44444444
[r]   Pos X offset (256x) 14: COLOR=44ffffff, left=144, 44efefff, top=143, 44efefef
[r]   Pos X offset 15: COLOR=440f0f0f, left=1, 4400ef0f, top=1, 44444444
[r]   Pos X offset (256x) 15: COLOR=44ffffff, left=144, 44efefff, top=143, 44efefef

[x] Pos Y offsets:
[r]   Pos Y offset 0: COLOR=44ffffff, left=1, 440000ff, top=1, 44000000
[r]   Pos Y offset (256x) 0: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos Y offset 1: COLOR=44efefff, left=1, 440000ff, top=1, 44000f00
[r]   Pos Y offset (256x) 1: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos Y offset 2: COLOR=44dfdfff, left=1, 440000ff, top=1, 44001f00
[r]   Pos Y offset (256x) 2: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos Y offset 3: COLOR=44cfcfff, left=1, 440000ff, top=1, 44002f00
[r]   Pos Y offset (256x) 3: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos Y offset 4: COLOR=44bfbfff, left=1, 440000ff, top=1, 44003f00
[r]   Pos Y offset (256x) 4: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos Y offset 5: COLOR=44afafff, left=1, 440000ff, top=1, 44004f00
[r]   Pos Y offset (256x) 5: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos Y offset 6: COLOR=449f9fff, left=1, 440000ff, top=1, 44005f00
[r]   Pos Y offset (256x) 6: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos Y offset 7: COLOR=448f8fff, left=1, 440000ff, top=1, 44006f00
[r]   Pos Y offset (256x) 7: COLOR=44ffffff, left=144, 44efefff, top=144, 44efefef
[r]   Pos Y offset 8: COLOR=447f7fff, left=1, 44444444, top=1, 44007f00
[r]   Pos Y offset (256x) 8: COLOR=44ffffff, left=143, 44efefff, top=144, 44efefef
[r]   Pos Y offset 9: COLOR=446f6fff, left=1, 44444444, top=1, 44008f00
[r]   Pos Y offset (256x) 9: COLOR=44ffffff, left=143, 44efefff, top=144, 44efefef
[r]   Pos Y offset 10: COLOR=445f5fff, left=1, 44444444, top=1, 44009f00
[r]   Pos Y offset (256x) 10: COLOR=44ffffff, left=143, 44efefff, top=144, 44efefef
[r]   Pos Y offset 11: COLOR=444f4fff, left=1, 44444444, top=1, 4400af00
[r]   Pos Y offset (256x) 11: COLOR=44ffffff, left=143, 44efefff, top=144, 44efefef
[r]   Pos Y offset 12: COLOR=443f3fff, left=1, 44444444, top=1, 4400bf00
[r]   Pos Y offset (256x) 12: COLOR=44ffffff, left=143, 44efefff, top=144, 44efefef
[r]   Pos Y offset 13: COLOR=442f2fff, left=1, 44444444, top=1, 4400cf00
[r]   Pos Y offset (256x) 13: COLOR=44ffffff, left=143, 44efefff, top=144, 44efefef
[r]   Pos Y offset 14: COLOR=441f1fff, left=1, 44444444, top=1, 4400df00
[r]   Pos Y offset (256x) 14: COLOR=44ffffff, left=143, 44efefff, top=144, 44efefef
[r]   Pos Y offset 15: COLOR=440f0fff, left=1, 44444444, top=1, 4400ef00
[r]   Pos Y offset (256x) 15: COLOR=44ffffff, left=143, 44efefff, top=144, 44efefef
